lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -81,8 +81,29 @@ def add_node_data(node)
     node.add_server_facts(@server_facts)
   end
 
+  # Inline file metadata for static catalogs
+  # Initially restricted to files sourced from codedir via puppet:/// uri.
+  def inline_metadata(catalog, environment)
+    catalog.resources.find_all { |res| res.type == "File" }.each do |resource|
+      next if resource[:ensure] == 'absent'
+
+      next unless source = resource[:source]
+      next unless source =~ /^puppet:/
+
+      file = resource.to_ral
+
+      if file.recurse?
+        # TODO: find and replace metadata for children recursively
+      else
+        # TODO: find and replace metadata for the file
+      end
+    end
+
+    catalog
+  end
```

Comment:
Would this logic be easier to test if it was moved to a different class, e.g. StaticCatalog{Filter,Inliner,Transformer,...} or some such, and then have the compiler terminus delegate to it, e.g. `catalog = StaticCatalogFilter.new(self, node).filter`?

Related, the `compiler` terminus already has some filtering logic to remove virtual resources, but the filter method gets called after `find` returns the catalog. I'm wondering if we should filter in a consistent way, either all the filtering happens in the `filter` method or in the `compile` method? 

Then we could have a base filter class that removes virtual resources, and have a static catalog filter class that extends it and also removes the resources above? Or we could chain filters?